goto/the_real_goto.py

```diff
@@ -86,6 +87,24 @@ def run_command(magic, command, args):
     if command in ['mv', 'rename']:
         return commands.rename(magic, command, args)
 
+    if command == 'use':
+        return commands.use(magic, command, args)
+
+    if command == 'project':
+        return None, GotoWarning('not_implemented')
+
+    if command == 'project:rm':
```

Comment:
very interesting

Reply:
Also a valid approach
`goto project rm <name>`
`goto project [project commands]`

Reply:
I think you approach is better. It is the intuitive thing to do.

Also I think that the use-command is the most important command. It is what the user needs 99% of the time from the project api, so it should be included as a first class command, and not hidden behind the project prefix.

```
goto use existing-project-name
goto use new-project-name
goto use https://gotohub.com/technocake/goto.goto

Reply:
Tl; DR - this is a topic we should discuss some more :) 

I think you present a set of very good ideas here. And I have some thoughts too.

First, the semantics should be clear and consistent - It should be clear to the user that it is projects  the user is working with.  I don't believe the semantics of `using` is clear enough. 


`goto use existing-project-name` vs `goto project existing-project-name` is almost the same. The first one is a bit shorter, and therefore better to type. But lacks clarity. The second one is more clear on what is happening.  Even for new users who haven't tried goto yet. 
Although the output of both commands would reveal that the project was switched.

In any case, the standalone `project` command is here to stay, regardless of what we let goto do with the project context directly. 

`goto use existing-project-name` -- this one might go fine, as long as it is a alias to `goto project existing-project-name` and the standalone `project existing-project-name`

`goto use new-project-name` - This behaviour I am actually against. I might open for it to be configurable.  But creating a new project should be a explicit command, and switching to a non-existent project should produce an error.
In the early revisions of goto, the `project <new-project>` behaved like proposed here. And the effect was many new misspelled projects, containing one shortcut here, and one shortcut there. 
It is also better for testability and if one wants to use goto in a scripting context if the commands are well behaved and exits 1 on switching to non-existing projects.

When working with python tools, like virtualenvwrapper - the creation differs in command from the activation - and activation of non-existing virtual environments yields error.  

The last two examples:
`goto use https://gotohub.com/technocake/remote-project-name.json`
`goto use alias-for-remote https://gotohub.com/g/RkwXwaZw2vQSpFcwX2s9.json`
are very cool. 

For these use cases, in the `shared context` I think the term `use` is supreme. 

One little detail - the concept of having all shortcuts in a single json file is soon to be a thing of the past. There will now be a `shared` and a `private` json file.

Reply:
On second thought I think having alias for remote after the url is better
```
goto use https://gotohub.com/g/RkwXwaZw2vQSpFcwX2s9.json alias-for-remote
```

I agree that having separate commands for switching and creation is a very good idea.